framework/include/userobject/UserObjectInterface.h

```diff
@@ -80,7 +83,18 @@ const T &
 UserObjectInterface::getUserObject(const std::string & name) const
 {
   unsigned int tid = needThreadedCopy(getUserObjectBase(name)) ? _uoi_tid : 0;
-  return _uoi_feproblem.getUserObject<T>(_uoi_params.get<UserObjectName>(name), tid);
+  const auto * uo = &_uoi_feproblem.getUserObject<T>(_uoi_params.get<UserObjectName>(name), tid);
+
+  if (!dynamic_cast<const T *>(uo))
+    _moose_object->paramError(name,
+                              "UserObject '" + _uoi_params.get<UserObjectName>(name) +
+                                  "' does not derive from " +
+                                  "the base class that is expected by " + _moose_object->type() +
+                                  ".\nPlease consult "
+                                  "the documentation for " +
+                                  _moose_object->type() + ".");
```

Comment:
```suggestion
                              "The supplied UserObject of type ", uo->type(), " does not derive from the base class expected (", typeid(T).name(), ")."
```

How about something like this?

Reply:
Unfortunately `typeid` is [implementation-dependent](https://en.cppreference.com/w/cpp/types/type_info/name), so what that prints can vary from useful to not-useful. For me, it prints:

```
P26SinglePhaseFluidProperties
```

for the test I added, where the 26 I've gleaned is the number of characters in the type name, and idk what P is.

Reply:
Ah yeah, I remember that. I typed this up without looking at any of the doco. I swore we had an implementation somewhere...